apps/sms/views/shared/js/navigation.js

```diff
@@ -1,361 +1,921 @@
+     from.panel.behavior !== newView.behavior);
+
+  if (!shouldPrepare) {
+    return;
+  }
 
-    if (nextPanelObject.beforeEnter) {
-      promise = promise.then(
-        nextPanelObject.beforeEnter.bind(nextPanelObject, transitionArgs)
+  var newPanelElement = document.querySelector(`.panel-${newView.behavior}`);
+  if (!newPanelElement) {
+    throw new Error(
+      'Couldn\'t find the container element for view ' + newView.name
+    );
+  }
+
+  newPanelElement.classList.add('panel-prepare-activation');
+}
+
+/**
+ * Hides current panel, shows new panel.
+ * It will use an animation if we change panels in the same document.
+ *
+ * @returns {Promise} Resolves when the animation is over, if there is an
+ * animation. Otherwise returns a resolved promise.
+ */
+function switchPanel() {
+  debug('switchPanel()');
+  if (!navigationTransition) {
+    return Promise.reject(new Error(
+      'Trying to switch panel but no transition is started yet.'
+    ));
+  }
+
+  var from = navigationTransition.states.from;
+  var to = navigationTransition.states.to;
+
+  var oldView;
+  var oldPanelElement;
+
+  if (from) {
+    oldView = from.panel;
+    oldPanelElement = document.querySelector(`.panel-${oldView.behavior}`);
+    if (!oldPanelElement) {
+      throw new Error(
+        'Couldn\'t find the container element for view ' + oldView.name
       );
     }
+  }
 
-    promise = promise.then(function resetPanel() {
-      // While we're sliding, we aren't in any panel
-      currentPanel = null;
-    });
+  var newView = to.panel;
+  var newPanelElement = document.querySelector(`.panel-${newView.behavior}`);
+  if (!newPanelElement) {
+    return Promise.reject(new Error(
+      'Couldn\'t find the container element for view ' + newView.name
+    ));
+  }
 
-    // sliding
-    promise = promise.then(
-      this.slide.bind(this, nextPanelInfo.wrapperPosition)
-    ).then(function changePanel() {
-      currentPanel = {
-        panel: panel,
-        args: args
-      };
-    });
+  var isGoingBack = oldView && oldView.previous === newView.name;
 
-    // afterLeave and afterEnter should not trigger errors
-    promise = promise.then(function() {
-      var promise = Promise.resolve();
+  newPanelElement.classList.remove('panel-prepare-activation');
 
-      if (currentPanelObject && currentPanelObject.afterLeave) {
-        promise = promise.then(
-          currentPanelObject.afterLeave.bind(currentPanelObject, transitionArgs)
-        ).catch(catchError);
-      }
+  var doSlideAnimation =
+    from && from.panel && newView &&
+    from.panel.behavior !== newView.behavior;
 
-      if (nextPanelObject.afterEnter) {
-        promise = promise.then(
-          nextPanelObject.afterEnter.bind(nextPanelObject, transitionArgs)
-        ).catch(catchError);
-      }
+  console.log('doSlideAnimation', doSlideAnimation);
+  if (doSlideAnimation) {
+    newPanelElement.classList.add('panel-will-activate', 'panel-active');
```

Comment:
it needs to be here otherwise it's not visible before the animation :)

I don't see it added later in the "doSlideAnimation" case ? (IMO this is more a case of "early returns make things more difficult to follow" and I should add a "else" block instead of early returning. What do you think ?)